File: plugins/orchestrator/skills/memory-bank/scripts/sync_metadata.py

import os
import json
import re
from datetime import datetime, timezone


def parse_frontmatter(content):
    """Parses YAML frontmatter from markdown content."""
    frontmatter_regex = r"^---\s+(.*?)\s+---"
    match = re.search(frontmatter_regex, content, re.DOTALL)
    if not match:
        return {}

    yaml_content = match.group(1)
    data = {}
    for line in yaml_content.split('\n'):
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        if ':' in line:
            key, value = line.split(':', 1)
            key = key.strip()
            value = value.strip()

            # Basic parsing for arrays [a, b]
            if value.startswith('[') and value.endswith(']'):
                value = [v.strip() for v in value[1:-1].split(',') if v.strip()]
            else:
                # Remove quotes if present
                if (value.startswith('"') and value.endswith('"')) or (value.startswith("'") and value.endswith("'")):
                    value = value[1:-1]

            data[key] = value
    return data


def collect_markdown_files(root_dir):
    files = []
    for dirpath, _, filenames in os.walk(root_dir):
        for filename in filenames:
            if filename.endswith('.md'):
                files.append(os.path.join(dirpath, filename))
    return files


def main():
    root_dir = os.getcwd()
    preferred_dir = os.environ.get('MEMORY_BANK_DIR') or os.environ.get('MEMORIES_DIR')
    candidates = (
        [os.path.join(root_dir, preferred_dir)]
        if preferred_dir
        else [os.path.join(root_dir, 'memory-bank'), os.path.join(root_dir, '.memories')]
    )

    memories_dir = ''
    for candidate in candidates:
        if os.path.isdir(candidate):
            memories_dir = candidate
            break

    if not memories_dir:
        print('Error: memory-bank or .memories directory not found.')
        return

    metadata_path = os.path.join(memories_dir, 'metadata.json')

    nodes = []
    edges = []

    for filepath in collect_markdown_files(memories_dir):
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()

        fm = parse_frontmatter(content)
        if 'id' not in fm:
            continue

        relative_path = os.path.relpath(filepath, memories_dir).replace('\\', '/')
        node = {
            'id': fm['id'],
            'file': relative_path,
            'type': fm.get('type', 'unknown'),
            'tags': fm.get('tags', []) if isinstance(fm.get('tags', []), list) else [],
        }
        nodes.append(node)

        links = fm.get('links', [])
        if isinstance(links, list):
            for target in links:
                edges.append({
                    'from': fm['id'],
                    'to': target,
                })

    nodes.sort(key=lambda item: item['id'])
    edges.sort(key=lambda item: (item['from'], item['to']))

    metadata = {
        'version': '1.0.0',
        'lastUpdated': datetime.now(timezone.utc).isoformat(),
        'nodes': nodes,
        'edges': edges,
    }

    with open(metadata_path, 'w', encoding='utf-8') as f:
        json.dump(metadata, f, indent=2)

    print(f"Synced {len(nodes)} nodes and {len(edges)} edges to {metadata_path}")


if __name__ == "__main__":
    main()

File: plugins/orchestrator/skills/memory-bank/scripts/sync_metadata.ts
import { existsSync, readdirSync, statSync } from 'node:fs';
import { join, relative } from 'node:path';

interface Frontmatter {
  id: string;
  type?: string;
  utc_datetime?: string;
  tags?: string[];
  links?: string[];
}

interface Node {
  id: string;
  file: string;
  type: string;
  tags: string[];
}

interface Edge {
  from: string;
  to: string;
}

interface Metadata {
  version: string;
  lastUpdated: string;
  nodes: Node[];
  edges: Edge[];
}

function parseFrontmatter(content: string): Frontmatter | null {
  // Extract YAML frontmatter between --- delimiters
  const frontmatterRegex = /^---\s+(.*?)\s+---/s;
  const match = content.match(frontmatterRegex);

  if (!match) {
    return null;
  }

  const yamlContent = match[1];

  try {
    // Use Bun's native YAML parser
    const data = Bun.YAML.parse(yamlContent) as Frontmatter;

    // Ensure we have the required fields
    if (!data.id) {
      return null;
    }

    return {
      id: data.id,
      type: data.type ?? 'unknown',
      utc_datetime: data.utc_datetime ?? '',
      tags: Array.isArray(data.tags) ? data.tags : [],
      links: Array.isArray(data.links) ? data.links : [],
    };
  } catch (error) {
    console.warn(`Warning: Failed to parse YAML frontmatter: ${error}`);
    return null;
  }
}

function collectMarkdownFiles(dir: string, results: string[]): void {
  const entries = readdirSync(dir, { withFileTypes: true });

  for (const entry of entries) {
    const entryPath = join(dir, entry.name);

    if (entry.isDirectory()) {
      collectMarkdownFiles(entryPath, results);
      continue;
    }

    if (entry.isFile() && entry.name.endsWith('.md')) {
      results.push(entryPath);
    }
  }
}

async function main(): Promise<void> {
  const rootDir = process.cwd();
  const preferredDir = process.env.MEMORY_BANK_DIR ?? process.env.MEMORIES_DIR;
  const candidates = preferredDir
    ? [join(rootDir, preferredDir)]
    : [join(rootDir, 'memory-bank'), join(rootDir, '.memories')];
  let memoriesDir = '';

  for (const candidate of candidates) {
    if (existsSync(candidate) && statSync(candidate).isDirectory()) {
      memoriesDir = candidate;
      break;
    }
  }

  if (!memoriesDir) {
    console.log('Error: memory-bank or .memories directory not found.');
    return;
  }

  const metadataPath = join(memoriesDir, 'metadata.json');

  const nodes: Node[] = [];
  const edges: Edge[] = [];

  const markdownFiles: string[] = [];
  collectMarkdownFiles(memoriesDir, markdownFiles);

  for (const filepath of markdownFiles) {
    const file = Bun.file(filepath);
    const content = await file.text();

    const fm = parseFrontmatter(content);

    if (fm) {
      const relativePath = relative(memoriesDir, filepath).split('\\').join('/');
      const node: Node = {
        id: fm.id,
        file: relativePath,
        type: fm.type,
        tags: fm.tags,
      };
      nodes.push(node);

      // Check for links and create edges
      const links = fm.links ?? [];
      for (const target of links) {
        edges.push({
          from: fm.id,
          to: target,
        });
      }
    }
  }

  nodes.sort((left, right) => left.id.localeCompare(right.id));
  edges.sort((left, right) => {
    const fromCompare = left.from.localeCompare(right.from);
    return fromCompare !== 0 ? fromCompare : left.to.localeCompare(right.to);
  });

  const metadata: Metadata = {
    version: '1.0.0',
    lastUpdated: new Date().toISOString(),
    nodes,
    edges,
  };

  // Write metadata.json
  await Bun.write(metadataPath, JSON.stringify(metadata, null, 2));

  console.log(`Synced ${nodes.length} nodes and ${edges.length} edges to ${metadataPath}`);
}

main().catch((error) => {
  console.error('Error:', error);
  process.exit(1);
});